Tools/build/check_warnings.py

```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env python3
+"""
+Parses compiler output with -fdiagnostics-format=json and checks that warnings exist
+only in files that are expected to have warnings.
+"""
+import argparse
+import json
+import re
+import sys
+from pathlib import Path
+
+def extract_warnings_from_compiler_output(compiler_output: str) -> list[dict]:
+    """
+    Extracts warnings from the compiler output when using -fdiagnostics-format=json
+
+    Compiler output as a whole is not a valid json document, but includes many json
+    objects and may include other output that is not json.
+    """
+
+    # Regex to find json arrays at the top level of the file in the compiler output
+    json_arrays = re.findall(r'\[(?:[^\[\]]|\[(?:[^\[\]]|\[[^\[\]]*\])*\])*\]', compiler_output)
+    compiler_warnings = []
+    for array in json_arrays:
+        try:
+            json_data = json.loads(array)
+            json_objects_in_array = [entry for entry in json_data]
+            compiler_warnings.extend([entry for entry in json_objects_in_array
+                                       if entry.get('kind') == 'warning'])
+        except json.JSONDecodeError:
+            continue  # Skip malformed JSON
```

Comment:
If we have a stricter regex like above, do we need to keep this step?

Reply:
> re.findall(r"^(\[{\"kind\".+\])$", compiler_output, flags=re.MULTILINE|re.DOTALL)
This regex looks like it requires the json arrays to be on a single line and gcc outputs the json arrays across multiple lines.

Were you able to extract the json arrays from the [sample gcc compiler output](https://github.com/python/cpython/pull/121730#discussion_r1680356865)

I am sure there is a more elegant way to grab them but for now the current regex successfully grabs all 156 warnings from the compiler output (many are duplicates)